### PR TITLE
Loki: Cache extracted labels

### LINKED ISSUE #75512: Loki autocomplete: too many sample requests
When we offer labels or parser suggestions, we grab the log query and request a log sample to extract them or infer them, in the case of the parser.

Even when the query didn't change and a sample was recently requested, we are sending a new sample query, resulting in too many queries in a short time. Additionally, each query has the potential to be slow to resolve in big environments, resulting in a degradation of the user experience.

https://github.com/grafana/grafana/assets/1069378/f90139dd-71bb-44d4-970c-133bf83469cc

### PR DESCRIPTION
**Pull request by gtk-grafana**
**What is this feature?**
Adds simple Map to cache the previous results of Loki data samples in the Loki language provider.

**Why do we need this feature?**
loki-data-samples queries are currently triggered directly from the Monaco completion callbacks, and not part of the react/UI application layer, so certain keystrokes in certain contexts will trigger another query. When this query takes a long time to complete, the editor UX is very sluggish and difficult to work with. Since the values returned by this query are extracted label values, which are not expected to change second-to-second, the solution proposed here is to cache requests and serve "stale" labels to prevent needing to wait for an api request whenever the user presses spacebar or comma inside a stream selector.

TL;DR;
To prevent duplicate API calls getting labels in autocomplete, we cache 2 unique query strings and their associated labels.

**Who is this feature for?**
Users of Loki code editor.

**Which issue(s) does this PR fix?**:
Fixes https://github.com/grafana/grafana/issues/75512

**Special notes for your reviewer:**
This is the lightest implementation I could imagine in terms of CPU/memory consumption, worried about cases where this list of labels is quite long.

We could use LRU for the added feature of smarter cache purging (instead of purging the first inserted, we'd be purging the least recently used), but that's a lot of additional overhead, and the API between map and LRU are very similar, so swapping it out would be very easy. 

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts [hunk 1]
```diff
@@ -55,8 +58,33 @@ export class CompletionDataProvider {
...
+   * Here is a lightweight and simple cache to avoid calling the backend multiple times for the same query.
+   *
+   * @param logQuery
+   */
+  async getParserAndLabelKeys(logQuery: string): Promise<ExtractedLabelKeys> {
+    const EXTRACTED_LABEL_KEYS_MAX_CACHE_SIZE = 2;
+    if (this.queryToLabelKeysCache.has(logQuery)) {
+      // Asserting the type here because we know something is in the cache from the use of Map.has() above
+      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+      return this.queryToLabelKeysCache.get(logQuery) as ExtractedLabelKeys;
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts:75]: fix

File: public/app/plugins/datasource/loki/types.ts [hunk 2]
```diff
@@ -84,4 +84,12 @@ export interface ContextFilter {
   description?: string;
 }
 
+export interface ExtractedLabelKeys {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/types.ts:87]: Yay for adding this 🤩!

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts [hunk 3]
```diff
@@ -16,7 +16,10 @@ export class CompletionDataProvider {
   constructor(
     private languageProvider: LanguageProvider,
     private historyRef: HistoryRef = { current: [] }
-  ) {}
+  ) {
+    this.queryToLabelKeysCache = new Map();
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts:20]: Just curious about why did you decide to use map and handle deleting/adding, rather than using `LRUCache` that we use in `LokiLanguageProvider` that handles additional logic for us, and we just need to `set` and `get`?

**Review by ivanahuckova — COMMENTED** [review]
Left a question bellow, but looks good. Great change!

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts (hunk 3 above)
Comment (gtk-grafana) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts:20]: We do already use the LRU cache, but in https://github.com/grafana/grafana/pull/75760 I'm proposing to replace it with the browser cache, so I didn't want to go all-in on LRU in case we want to get away from it.

If we're ok using a bit more memory we could certainly use LRU and/or increase the size of the cache, for this MVP I was going for the fastest/lightest possible implementation.

Swapping out for LRU should be trivial as it has the same interface as Map, we'll just need to delete the bit where we manually delete the oldest element and swap out the Map initializer for LRUCache

Comment (gtk-grafana) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts:20]: I guess the question is how many items do we want to cache, and how long do we want to cache them, and how bad are stale records being served throughout the session?

I guess the question becomes: is the goal of this cache to prevent duplicate requests, or to cache every possible request in a session?
If we want to cache "everything" or as much as possible, we should use LRU, or beef up the size of the map.
If we're worried about stale results we should range snap and check the date ranges when checking the cache (although in the `getParserAndLabelKeys` we don't have the time range in scope), or stick to a small cache size so results aren't likely to stick around for a long time.

LRU is ~49K, which IMO is a tad overkill for a map interface that keeps track of how many times each element is hit.

**Review by ivanahuckova — APPROVED** [review]
LGTM! Nice change 👍
